File: library/src/main/java/io/github/anderscheow/validator/rules/common/MaxRule.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import java.util.Locale;

import io.github.anderscheow.validator.rules.BaseRule;

public class MaxRule extends BaseRule {

    private int maxLength;

    public MaxRule(int maxLength) {
        super(String.format(Locale.getDefault(), "Length must not exceed %d characters", maxLength));
        this.maxLength = maxLength;
    }

    public MaxRule(int maxLength, @StringRes int errorRes) {
        super(errorRes);
        this.maxLength = maxLength;
    }

    public MaxRule(int maxLength, @NonNull String errorMessage) {
        super(errorMessage);
        this.maxLength = maxLength;
    }

    @Override
    public boolean validate(Object value) {
        if (value == null) {
            throw new NullPointerException();
        }

        if (value instanceof String) {
            return ((String) value).length() <= maxLength;
        }

        throw new ClassCastException("Required String value");
    }
}

File: library/src/main/java/io/github/anderscheow/validator/rules/common/MaxRule.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import io.github.anderscheow.validator.rules.Rule

class MaxRule : Rule {

    private var maxLength: Int

    constructor(maxLength: Int, @StringRes errorRes: Int) :
            super(errorRes) {
        this.maxLength = maxLength
    }

    constructor(maxLength: Int, errorMessage: String) :
            super(errorMessage) {
        this.maxLength = maxLength
    }

    override fun validate(value: String?): Boolean {
        if (value == null) {
            throw NullPointerException()
        } else {
            return value.length <= maxLength
        }
    }
}

fun maximumLength(maxLength: Int, @StringRes errorRes: Int): MaxRule = MaxRule(maxLength, errorRes)

fun maximumLength(maxLength: Int, errorMessage: String): MaxRule = MaxRule(maxLength, errorMessage)
